sweep/fee_bumper.go

```diff
@@ -0,0 +1,894 @@
+	defer t.wg.Done()
+
+	// Create a result that will be sent to the resultChan which is
+	// listened by the caller.
+	result := &BumpResult{
+		Event:     TxConfirmed,
+		Tx:        tx,
+		requestID: requestID,
+	}
+
+	// Notify that this tx is confirmed and remove the record from the map.
+	t.handleResult(result)
+}
+
+// handleFeeBumpTx checks if the tx needs to be bumped, and if so, it will
+// attempt to bump the fee of the tx.
+//
+// NOTE: Must be run as a goroutine to avoid blocking on sending the result.
+func (t *TxPublisher) handleFeeBumpTx(requestID uint64, r *monitorRecord) {
+	defer t.wg.Done()
+
+	oldTxid := r.tx.TxHash()
+
+	// Get the current conf target for this record.
+	confTarget := calcCurrentConfTarget(
+		t.currentHeight, r.req.DeadlineHeight,
+	)
+
+	// Ask the fee function whether a bump is needed. We expect the fee
+	// function to increase its returned fee rate after calling this
+	// method.
+	if r.feeFunction.SkipFeeBump(confTarget) {
+		log.Debugf("Skip bumping tx %v at height=%v", oldTxid,
+			t.currentHeight)
+
+		return
+	}
+
+	// The fee function now has a new fee rate, we will use it to bump the
+	// fee of the tx.
+	result, err := t.createAndPublishTx(requestID, r)
+	if err != nil {
+		log.Errorf("Failed to bump tx %v: %v", oldTxid, err)
+
+		return
+	}
+
+	// Notify the new result.
+	t.handleResult(result)
+}
+
+// createAndPublishTx creates a new tx with a higher fee rate and publishes it
+// to the network. It will update the record with the new tx and fee rate if
+// successfully created, and return the result when published successfully.
+func (t *TxPublisher) createAndPublishTx(requestID uint64,
+	r *monitorRecord) (*BumpResult, error) {
+
+	// Fetch the old tx.
+	oldTx := r.tx
+
+	// Create a new tx with the new fee rate.
+	//
+	// NOTE: The fee function is expected to have increased its returned
+	// fee rate after calling the SkipFeeBump method. So we can use it
+	// directly here.
+	tx, fee, err := t.createAndCheckTx(r.req, r.feeFunction)
+
+	// If the tx doesn't not have enought budget, we will return a result
+	// so the sweeper can handle it by re-clustering the utxos.
+	if errors.Is(err, ErrNotEnoughBudget) {
+		log.Warnf("Fail to fee bump tx %v: %v", oldTx.TxHash(), err)
+
+		return &BumpResult{
+			Event:     TxFailed,
+			Tx:        oldTx,
+			Err:       err,
+			requestID: requestID,
+		}, nil
+	}
+
+	// If the error is not budget related, we will return an error and let
+	// the fee bumper retry it at next block.
+	//
+	// NOTE: we can check the RBF error here and ask the fee function to
+	// recalculate the fee rate. However, this would defeat the purpose of
+	// using a deadline based fee function:
+	// - if the deadline is far away, there's no rush to RBF the tx.
+	// - if the deadline is close, we expect the fee function to give us a
+	//   higher fee rate. If the fee rate cannot satisfy the RBF rules, it
+	//   means the budget is not enough.
+	if err != nil {
```

Comment:
yeah the wallet utxo is a headache...I'm thinking maybe we could use `LeaseOutput` to lock it for a while? Then it'd still have the issue, as you mentioned in the comments, that we may have fewer utxos available for funding - guess it's a tradeoff we have to make here.

Reply:
You know, it occurs to me that using wallet outputs at all may be neither necessary nor good. If wallet inputs are needed doesn't that imply that all of the swept outputs are negative yielding? Why would we want to pull in additional wallet inputs?

EDIT: I suspect it's because we can't alter the output values for SS-HTLC txs due to SIGHASH_SINGLE, meaning we'd like to reduce the output value to create more fee room but it would nuke our signature. Can one of you confirm this?

Reply:
> If wallet inputs are needed doesn't that imply that all of the swept outputs are negative yielding? Why would we want to pull in additional wallet inputs?

One case wallet inputs are needed is when we're sweeping anchor outputs. They're only 300 or so satoshis, so wallet inputs are always required to bump the package. 

Another scenario is when a user is attempting to use `BumpFee` to increase the fee of a wallet transaction. 

A final (and the most fundamental case) is that for second level HTLCs, the second level transaction actually has zero fees applied. As a result, any time you want to broadcast such a transaction, we need to pull in additional wallet inputs.